Exception table support

82marbag opened the pull request:
*Issue #, if available:* #6

*Description of changes:*

This PR adds support for exception tables and uses them to implement rdmsr_safe and wrmsr_safe, that will fail by reporting the error in a boolean variable if the operation on an msr can't be completed due to an exception.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

82marbag committed:
#6 support exception tables

Add support for exception tables.
It is possible to register an address that triggers
an exception to call an handler, jump to a fixup address
or both.

Signed-off-by: Daniele Ahmed <ahmeddan@amazon.com>

wipawel commented on arch/x86/traps.c:
```diff
@@ -289,6 +291,22 @@ void print_callstack(const void *sp, const void *ip) {
 
 void do_exception(struct cpu_regs *regs) {
     static char ec_str[32], panic_str[128];
+    for (extable_entry_t *cur = __start_extables; cur < __end_extables; ++cur) {
```
nit: I would put this code into a separate static inline function

wipawel commented on include/extables.h:
```diff
@@ -0,0 +1,45 @@
+
+#include <arch/x86/processor.h>
+
+typedef void(extable_entry_callback_t)(cpu_regs_t *regs);
```
`typedef void (*extable_entry_callback_t)(cpu_regs_t *regs);`

wipawel commented on include/extables.h:
```diff
@@ -0,0 +1,45 @@
+
+struct extable_entry {
+    uint64_t fault_addr;
+
```
nit: extra newline

wipawel commented on include/extables.h:
```diff
@@ -0,0 +1,45 @@
+typedef void(extable_entry_callback_t)(cpu_regs_t *regs);
+
+struct extable_entry {
+    uint64_t fault_addr;
+
+    uint64_t fixup;
```
maybe `void *` would be a better type?

82marbag replied:
Actually `x86_reg_t` would be a better one

wipawel commented on include/extables.h:
```diff
@@ -0,0 +1,45 @@
+
+    uint64_t fixup;
+    extable_entry_callback_t *cb;
+};
```
might need `__packed` directive

wipawel commented on include/extables.h:
```diff
@@ -0,0 +1,45 @@
+typedef struct extable_entry extable_entry_t;
+
+#define ASM_EXTABLE_HANDLER(fault_address, fixup, handler)                               \
+    ".pushsection .extables;\n"                                                          \
+    ".quad " STR(fault_address) ", " STR(fixup) ", " STR(handler) ";\n"                  \
+                                                                  ".popsection;\n"
```
nit: could you add generic pushsection/popsection macro `to asm_macros.h`?

wipawel commented on include/lib.h:
```diff
@@ -144,11 +145,42 @@ static inline uint64_t rdmsr(uint32_t msr_idx) {
     return (((uint64_t) high) << 32) | low;
 }
 
+/*
+ * read the msr_idx msr; the result is valid iff *read is true
+ */
+static inline uint64_t rdmsr_safe(uint32_t msr_idx, bool *read) {
```
nit: I would invert the logic here: the function can return `int` and set the value to input parameter. Same for `wrmsr_safe`.

wipawel replied:
Please make this and the following function a no inline functions. Otherwise there will be more exceptions registered than necessary.

wipawel commented on include/lib.h:
```diff
@@ -144,11 +145,42 @@ static inline uint64_t rdmsr(uint32_t msr_idx) {
+    volatile bool _read = false;
+    uint32_t low, high;
+
+    asm volatile("1: rdmsr; movq $1, %[read];"
```
this code could be made simpler with `asm goto` directive, the supplied in label can point to the error condition.

wipawel commented:
Please also fix in the `drivers/acpi/acpica/osl.c`:

```
/* FIXME: Check if pages are mapped (with exception tables) */
BOOLEAN AcpiOsReadable(void *Memory, ACPI_SIZE Length) { return true; }

/* FIXME: Check if pages are mapped and writeable (with exception tables) */
BOOLEAN AcpiOsWriteable(void *Memory, ACPI_SIZE Length) { return true; }
```

wipawel commented:
nit: please also add empty `init_extable()` and call it from `setup.c` replacing `    /* TODO: Exception tables */`

82marbag committed:
rdmsr and wrmsr _safe with exception tables

Add new rdmsr_safe and wrmsr_safe. These two functions
use exception tables to try performing their operations
and returning safely if the operation could not be performed.

Signed-off-by: Daniele Ahmed <ahmeddan@amazon.com>

82marbag committed:
acpica: AcpiOsReadable and AcpiOsWriteable with extables

In AcpiOsReadable and AcpiOsWriteable the checks for mapped pages
and their permissions for reading and writing respectively
are performed with exception tables

Signed-off-by: Daniele Ahmed <ahmeddan@amazon.com>

wipawel commented on arch/x86/traps.c:
```diff
@@ -287,9 +289,30 @@ void print_callstack(const void *sp, const void *ip) {
     printk("\n");
 }
 
+static void use_extables(struct cpu_regs *regs) {
```
nit: add inline

wipawel commented on drivers/acpi/acpica/osl.c:
```diff
@@ -207,11 +207,49 @@ void *AcpiOsAllocate(ACPI_SIZE Size) { return kmalloc(Size); }
+    for (size_t i = 0; i < Length / PAGE_SIZE; i += PAGE_SIZE) {
+        success = false;
+        asm volatile("movq %[mem], %%rax;"
+                     "1: testb $0, (%%rax); movq $1, %[success];"
```
you can just do: `movb (%[mem]), %%rax;`

wipawel commented on drivers/acpi/acpica/osl.c:
```diff
@@ -207,11 +207,49 @@ void *AcpiOsAllocate(ACPI_SIZE Size) { return kmalloc(Size); }
+    volatile bool success = false;
+    char *mem = (char *) Memory;
+
+    for (size_t i = 0; i < Length / PAGE_SIZE; i += PAGE_SIZE) {
```
this loop will not check final page for sizes different than multiple of PAGE_SIZE

wipawel commented on drivers/acpi/acpica/osl.c:
```diff
@@ -207,11 +207,49 @@ void *AcpiOsAllocate(ACPI_SIZE Size) { return kmalloc(Size); }
+        success = false;
+        asm volatile("movq %[mem], %%rax;"
+                     "1: testb $0, (%%rax); movq $1, %[success];"
+                     "2:" ASM_EXTABLE_HANDLER(1b, 2b, 0)
```
use ASM_EXTABLE()

wipawel commented on include/lib.h:
```diff
@@ -27,6 +27,7 @@
 
 #include <asm-macros.h>
 #include <console.h>
+#include <extables.h>
```
not needed?

wipawel commented on lib/lib.c:
```diff
@@ -42,3 +42,33 @@ int rand(void) {
+    uint32_t low, high;
+
+    asm volatile("1: rdmsr; movb $1, %[success];"
+                 "2:" ASM_EXTABLE_HANDLER(1b, 2b, 0)
```
use ASM_EXTABLE()

wipawel commented on lib/lib.c:
```diff
@@ -42,3 +42,33 @@ int rand(void) {
+    volatile bool success = false;
+
+    asm volatile("1: wrmsr; movq $1, %[success];"
+                 "2:" ASM_EXTABLE_HANDLER(1b, 2b, 0)
```
use ASM_EXTABLE()

wipawel commented on include/extables.h:
```diff
@@ -0,0 +1,47 @@
+
+#define ASM_EXTABLE_HANDLER(fault_address, fixup, handler)                               \
+    PUSHSECTION(.extables)                                                               \
+    ".quad " STR(fault_address) ", " STR(fixup) ", " STR(handler) ";\n" POPSECTION
```
there should also be a version for 32bit, where DWORD is used instead

wipawel replied:
or use uint64_t in `struct extable_entry`, but that would waste space on 32bit
